contracts/Curation.sol

```diff
@@ -0,0 +1,415 @@
+        require(canBeRegistered(entryId));
+
+        Application storage application = applications[entryId];
+        require(!application.registered);
+
+        // insert in Registry app
+        registry.add(application.data);
+        application.registered = true;
+    }
+
+    function setVotingApp(IVoting _voting) authP(CHANGE_VOTING_APP_ROLE, arr(voting, _voting)) public {
+        _setVotingApp(_voting);
+    }
+
+    function setMinDeposit(uint256 _minDeposit) authP(CHANGE_PARAMS_ROLE, arr(uint256(MIN_DEPOSIT_PARAM), _minDeposit)) public {
+        _setMinDeposit(_minDeposit);
+    }
+
+    function setApplyStageLen(uint64 _applyStageLen) authP(CHANGE_PARAMS_ROLE, arr(uint256(APPLY_STAGE_LEN_PARAM), _applyStageLen)) public {
+        _setApplyStageLen(_applyStageLen);
+    }
+
+    function setDispensationPct(uint256 _dispensationPct)
+        authP(CHANGE_PARAMS_ROLE, arr(uint256(DISPENSATION_PCT_PARAM), _dispensationPct))
+        public
+    {
+        require(_dispensationPct <= PCT_BASE);
+
+        _setDispensationPct(_dispensationPct);
+    }
+
+    function canBeRegistered(bytes32 entryId) view public returns (bool) {
+        // no challenges
+        if (getTimestamp() > applications[entryId].date.add(applyStageLen) &&
+            challenges[entryId].challenger == address(0))
+        {
+            return true;
+        }
+
+        return false;
+    }
+
+    function isChallengeResolved(bytes32 entryId) view public returns (bool) {
+        return challenges[entryId].resolved;
+    }
+
+    function getApplication(
+        bytes32 entryId
+    )
+        view
+        public
+        returns (
+            address applicant,
+            uint64 date,
+            bool registered,
+            bytes data,
+            uint256 amount,
+            uint256 lockId
+        )
+    {
+        Application memory application = applications[entryId];
+        return (
+            application.applicant,
+            application.date,
+            application.registered,
+            application.data,
+            application.amount,
+            application.lockId
+        );
+    }
+
+    function getChallenge(
+        bytes32 entryId
+    )
+        view
+        public
+        returns (
+            address challenger,
+            uint64 date,
+            bool resolved,
+            uint256 amount,
+            uint256 lockId,
+            uint256 voteId,
+            uint256 dipsensationPct
+        )
+    {
+        Challenge memory challenge = challenges[entryId];
+        return (
+            challenge.challenger,
+            challenge.date,
+            challenge.resolved,
+            challenge.amount,
+            challenge.lockId,
+            challenge.voteId,
+            challenge.dispensationPct
+        );
+    }
+
+    function _checkLock(address user, uint256 lockId, uint64 date) internal returns (uint256) {
+        // check lockId was not used before
+        require(!usedLocks[user][lockId]);
```

Comment:
Even though we can be fairly sure that the staking app is not malicious and it is not going to re-enter, we can have this check after the `staking.getLock(msg.sender, lockId)` to make sure the storage is correct after the external call.

Ideally we would do a `STATICCALL` but it is too cumbersome in this version of solidity.